java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoadingGood.java

```diff
@@ -0,0 +1,55 @@
+
+import android.app.Application;
+import android.content.Context;
+import android.content.pm.PackageInfo;
+import android.os.Bundle;
+
+public class InsecureClassLoading extends Application {
+  @Override
+  public void onCreate() {
+    super.onCreate();
+    invokePlugins();
+  }
+
+  private void invokePlugins() {
+
+    PackageManager packageManager = getPackageManager();
+    List<PackageInfo> packageList = packageManager.getInstalledPackages(PackageManager.GET_SIGNATURES);
+
+    for (PackageInfo p : packageList) {
+
+      // Check the signature of the package before using it.
+      final Signature[] arrSignatures = p.signatures;
+      for (final Signature sig : arrSignatures) {
+        final byte[] rawCert = sig.toByteArray();
+        InputStream certStream = new ByteArrayInputStream(rawCert);
+
+        try {
+          CertificateFactory certFactory = CertificateFactory.getInstance("X509");
+          X509Certificate x509Cert = (X509Certificate) certFactory.generateCertificate(certStream);
```

Comment:
To check: does this just get the presented certificate, or does it verify the chain of trust as well?

Reply:
Only the presented certificate I think.